### PR TITLE
stdlib: Improve handing of errors in Atlas request failures

### LINKED ISSUE #340: Retry, rather than immediately fail when `urlopen` Exception raised in "client_wrapper.py"
Fairly frequently tests will fail to download a resource. If you look into the stderr you see errors like the following:

```sh
warn: Error getting resources from client gem5-resources: <urlopen error [Errno -2] Name or service not known>
Exception: Resource with ID 'rv32ud-ps-fmin' not found.

At:
  src/python/gem5/resources/client_api/client_wrapper.py(151): get_resource_json_obj_from_client
  src/python/gem5/resources/client.py(152): get_resource_json_obj
  src/python/gem5/resources/resource.py(658): obtain_resource
  /__w/gem5/gem5/tests/gem5/asmtest/configs/riscv_asmtest.py(108): <module>
  src/python/m5/main.py(630): main
```

Based on the "Name or service not known", I believe in this case it's due to DNS lookup being unsuccessful. I don't know why this happens but I'd guess it's just a bad networking issue on one of the services we host our GitHub Runners (I suspect Amarillo... I've had problems with its networking before).

Regardless, in this case, and in others regarding networking issues, it'd be good if this code retried a few times before failing. This would make our tests considerably less flakey. This error has nothing to do with errors in our codebase, only errors in our networking (I _think_).

I'd suggest implementing a back-off algorithm that retries a few times with some idle time between each attempt. A niave approach could be : retry immediately; retried after 10 seconds; retried after 20 seconds; retry after 30 second. After this we can fail. `urlopen` may provide some functions to do this for you, I haven't bothered to look.

### PR DESCRIPTION
**Pull request by BobbyRBruce**
Now:

* The Atlas Client will attempt a connection 4 times, using an exponential backoff approach between attempts.
* When a failure does arise a rich output is given so problems can be easily diagnosed.

Addresses: #340

### CODE REVIEW

**Comment by BobbyRBruce** [comment]
@Harshil2107 : I was looking over #392 and had a few suggestions. This then spiraled into me just doing it. Open for feedback here. I haven't tested this yet (so pseudo-drafty), but I think the design idea is sound.

**Comment by BobbyRBruce** [comment]
I don't know what's going on, but the tests are failing with the 404 error, so at least we can see what the error would look like to the user:

```sh
warn: `Resource` has been deprecated. Please use the `obtain_resource` function instead.
warn: Attempt 1 of Atlas HTTP Request failed.
Purpose of Request: Get Access Token with API key.

Failed with Exception:
HTTP Error 404: Not Found

Retrying after 2 seconds...
warn: Attempt 2 of Atlas HTTP Request failed.
Purpose of Request: Get Access Token with API key.

Failed with Exception:
HTTP Error 404: Not Found

Retrying after 4 seconds...
warn: Attempt 3 of Atlas HTTP Request failed.
Purpose of Request: Get Access Token with API key.

Failed with Exception:
HTTP Error 404: Not Found

Retrying after 8 seconds...
Exception: Error getting resources from client gem5-resources

At:
  src/python/gem5/resources/client_api/client_wrapper.py(117): get_all_resources_by_id
  src/python/gem5/resources/client_api/client_wrapper.py(150): get_resource_json_obj_from_client
  src/python/gem5/resources/client.py(152): get_resource_json_obj
  src/python/gem5/resources/resource.py(736): obtain_resource
  src/python/gem5/resources/resource.py(975): Resource
  /__w/gem5/gem5/tests/gem5/m5_util/configs/simple_binary_run.py(92): <module>
  src/python/m5/main.py(642): main
```

This somehow isn't giving the errors information from the `AtlasClient` like i thought they'd be. This is a bug.

**Comment by BobbyRBruce** [comment]
Think i've fixed the issue partially in the latest push... we'll see.

File: src/python/gem5/resources/client_api/atlasclient.py [hunk 1]
```diff
@@ -47,22 +77,70 @@ def __init__(self, config: Dict[str, str]):
...
+            data_json={"key": self.apiKey},
+            headers={"Content-Type": "application/json"},
+            purpose_of_request="Get Access Token with API key",
+        )["access_token"]
+
+    def _atlas_http_json_req(
+        self,
+        url: str,
+        data_json: Dict[str, Any],
+        headers: dict[str, str],
```
Comment (Harshil2107) [inline src/python/gem5/resources/client_api/atlasclient.py]: Shouldn't this be `Dict[str, str]`?

File: src/python/gem5/resources/client_api/atlasclient.py [hunk 2]
```diff
@@ -47,22 +77,70 @@ def __init__(self, config: Dict[str, str]):
...
+        for attempt in itertools.count(start=1):
+            try:
+                response = request.urlopen(req)
+                break
+            except Exception as e:
+                if attempt >= max_failed_attempts:
+                    raise AtlasClientHttpJsonRequestError(
+                        client=self,
+                        data=data_json,
+                        headers=headers,
```
Comment (Harshil2107) [inline src/python/gem5/resources/client_api/atlasclient.py]: `header` is being passed as a parameter to `AtlasClientHttpJsonRequestError` but it does not have a parameter named `header`.

**Review by powerjg — APPROVED** [review]
lgtm

**Pull request merged**